File: client/src/domain/__tests__/sapOrderState.test.ts
/**
 * SAP's view of an order, turned into something the screens act on.
 *
 * `shared/fixtures/sap_order_state.json` states the rule; this asserts the
 * dashboard obeys it and `app/test/sap_order_state_test.dart` asserts the
 * phone does. Both read the same cases, because a rep and a manager looking
 * at the same order must be told the same thing about it.
 */

import { describe, expect, it } from 'vitest';
import cases from '../../../../shared/fixtures/sap_order_state.json';
import {
  cancelledInSap,
  orderBucket,
  orderPill,
  orderProgress,
  ORDER_BUCKET_LABEL,
  lineStatusFromSap,
  orderStatusFromLines,
  productionStatusFromSap,
  reachedSap,
  sapStale,
  sapSummary,
} from '../sapOrderState';

/** A fixture's raw ERPNext document, read the way `toOrderDetail` reads one. */
function fromDoc(o: Record<string, string | undefined>) {
  return {
    salesOrder: o.custom_sap_sales_order,
    salesOrderStatus: o.custom_sap_sales_order_status,
    invoice: o.custom_sap_invoice,
    invoiceDate: o.custom_sap_invoice_date,
  };
}

describe('an order', () => {
  for (const c of cases.order_status) {
    it(c.why, () => {
      expect(productionStatusFromSap({ salesOrder: c.sap_order, invoice: c.invoice })).toBe(c.expect);
    });
  }
});

describe('one line of an order', () => {
  for (const c of cases.line_status) {
    it(c.why, () => {
      expect(lineStatusFromSap({ invoice: c.line_invoice }, { salesOrder: c.order_sap })).toBe(
        c.expect,
      );
    });
  }

  it('without its order, an uninvoiced line cannot claim to be in SAP', () => {
    expect(lineStatusFromSap({})).toBe('Not Started');
  });
});

describe('what an order list shows', () => {
  for (const c of cases.order_progress) {
    it(c.why, () => {
      expect(orderProgress({ salesOrder: c.sap_order, invoice: c.invoice }, c.stored)).toBe(c.expect);
    });
  }

  it('a missing stored status is Not Started', () => {
    expect(orderProgress({}, undefined)).toBe('Not Started');
  });
});

describe('an order rolls up from its lines', () => {
  for (const c of cases.order_rolls_up_from_lines as {
    why: string;
    order_sap: string;
    order_invoice?: string;
    line_invoices: string[];
    expect: string;
  }[]) {
    it(c.why, () => {
      const lines = c.line_invoices.map((invoice) => ({ invoice }));
      expect(
        orderStatusFromLines(lines, { salesOrder: c.order_sap, invoice: c.order_invoice }),
      ).toBe(c.expect);
    });
  }
});

describe('the one line a rep reads', () => {
  for (const c of cases.summary as {
    why: string;
    order: Record<string, string | undefined>;
    expect: string | null;
  }[]) {
    it(c.why, () => {
      expect(sapSummary(fromDoc(c.order))).toBe(c.expect);
    });
  }
});

describe('the mistakes this mapping exists to prevent', () => {
  it("Frappe's string 'null' is not an invoice", () => {
    // An unset field read back through naive interpolation. Treating it as an
    // invoice would mark an order as gone that has not left.
    expect(lineStatusFromSap({ invoice: 'null' }, { salesOrder: '412' })).toBe('Pushed to SAP');
  });

  it('a stale production stage from before 24 Sep 2026 is never read', () => {
    // Fields the type no longer carries cannot reach the derivation at all;
    // this pins that an order carrying one is judged on its invoice alone.
    const doc = { custom_sap_sales_order: '412', custom_sap_production_stage: 'Closed' };
    expect(productionStatusFromSap(fromDoc(doc))).toBe('Pushed to SAP');
  });

  it('not yet in SAP is not the same as failed', () => {
    expect(reachedSap({})).toBe(false);
    expect(reachedSap({ salesOrder: 'SO-1' })).toBe(true);
    expect(reachedSap({ salesOrder: '   ' })).toBe(false);
  });
});

describe('staleness', () => {
  const now = new Date('2026-09-12T10:00:00');

  it('an order SAP never took is not stale, it is waiting', () => {
    expect(sapStale({}, now)).toBe(false);
  });

  it('in SAP but never reconciled is stale', () => {
    expect(sapStale({ salesOrder: 'SO-1' }, now)).toBe(true);
  });

  it('reconciled this morning is not stale', () => {
    expect(sapStale({ salesOrder: 'SO-1', syncedAt: '2026-09-12T08:00:00' }, now)).toBe(false);
  });

  it('reconciled two days ago is stale', () => {
    expect(sapStale({ salesOrder: 'SO-1', syncedAt: '2026-09-10T08:00:00' }, now)).toBe(true);
  });
});

describe('an order SAP has cancelled', () => {
  /*
   * The failure this closes, found 18 September 2026: SAP order 399 had been
   * cancelled and the sync had recorded `bost_Cancelled` on the ERPNext order
   * correctly, for days. Nothing read it, so the sales manager's board went on
   * showing the order as approved.
   */
  for (const c of cases.cancelled_orders as {
    why: string;
    order: Record<string, string>;
    expect_cancelled: boolean;
  }[]) {
    it(c.why, () => {
      expect(
        cancelledInSap({ salesOrderStatus: c.order.custom_sap_sales_order_status }),
      ).toBe(c.expect_cancelled);
    });
  }

  it('says CANCELLED IN SAP however the approval status reads', () => {
    const pill = orderPill('PO Approved - Ready for SAP', {
      salesOrderStatus: 'bost_Cancelled',
    });
    expect(pill.text).toBe('CANCELLED IN SAP');
    expect(pill.tone).toBe('danger');
  });

  it('leaves every other order showing its approval status', () => {
    expect(orderPill('PO Approved - Ready for SAP', { salesOrderStatus: 'bost_Open' }).text).toBe(
      'APPROVED',
    );
    expect(orderPill('Pending Approval', {}).text).toBe('WAITING FOR MANAGER APPROVAL');
    // Closed is finished, not cancelled. Conflating them would retire a
    // delivered order as though it had been called off.
    expect(orderPill('PO Approved - Ready for SAP', { salesOrderStatus: 'bost_Close' }).text).toBe(
      'APPROVED',
    );
  });
});

describe('the four buckets a manager filters their team orders by', () => {
  it('an order waiting on the manager is To approve', () => {
    expect(orderBucket({ approved: false, poStatus: 'Pending Approval' })).toBe('to_approve');
  });

  it('one escalated to the GM is still To approve — somebody owes a decision', () => {
    expect(orderBucket({ approved: false, poStatus: 'Pending GM Approval' })).toBe('to_approve');
  });

  it('an approved order is Approved', () => {
    expect(
      orderBucket({ approved: true, poStatus: 'PO Approved - Ready for SAP' }),
    ).toBe('approved');
  });

  it('a rejected order is Rejected, not To approve', () => {
    // awaitingManager() counts Rejected as still owing a decision, which is
    // right for the header count and wrong for this filter: a manager picking
    // "To approve" wants what they can act on now.
    expect(orderBucket({ approved: false, poStatus: 'Rejected' })).toBe('rejected');
  });

  /*
   * The trap this function exists for. A cancelled order still carries
   * custom_po_status = 'PO Approved - Ready for SAP', so testing approval
   * first files it under Approved while its own pill reads CANCELLED IN SAP —
   * two answers to the same question on one screen.
   */
  it('cancellation outranks approval, exactly as the pill does', () => {
    const cancelled = {
      approved: true,
      poStatus: 'PO Approved - Ready for SAP',
      salesOrderStatus: 'bost_Cancelled',
    };
    expect(orderBucket(cancelled)).toBe('cancelled');
    expect(orderPill(cancelled.poStatus, { salesOrderStatus: cancelled.salesOrderStatus }).text).toBe(
      'CANCELLED IN SAP',
    );
  });

  it('cancellation outranks rejection too', () => {
    expect(
      orderBucket({ approved: false, poStatus: 'Rejected', salesOrderStatus: 'bost_Cancelled' }),
    ).toBe('cancelled');
  });

  it('closed is finished, not cancelled — it stays Approved', () => {
    expect(
      orderBucket({
        approved: true,
        poStatus: 'PO Approved - Ready for SAP',
        salesOrderStatus: 'bost_Close',
      }),
    ).toBe('approved');
  });

  it('an order SAP has not seen is bucketed on its approval status alone', () => {
    expect(orderBucket({ approved: true, poStatus: 'PO Approved - Ready for SAP' })).toBe(
      'approved',
    );
    expect(orderBucket({ approved: false })).toBe('to_approve');
  });

  it("a lead order counting as approved lands in Approved, though isApproved would refuse it", () => {
    // Lead Order.status = 'Approved' is not PO_STATUS.approved. The caller
    // resolves that with leadOrderApproved and passes the answer in.
    expect(orderBucket({ approved: true, poStatus: 'Approved' })).toBe('approved');
    expect(orderBucket({ approved: true, poStatus: 'Converted' })).toBe('approved');
  });

  it('every bucket has a label, so a filter can never render undefined', () => {
    for (const b of ['to_approve', 'approved', 'rejected', 'cancelled'] as const) {
      expect(ORDER_BUCKET_LABEL[b]).toBeTruthy();
    }
  });
});

File: app/test/sap_order_state_test.dart
// SAP's view of an order, turned into something the screens act on.
//
// Read from the shared fixture the dashboard's suite reads too: a rep and a
// manager looking at the same order must be told the same thing about it.

import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

import 'package:manna_field_sales/core/sap_order_state.dart';

List<Map<String, dynamic>> _cases(Map<String, dynamic> f, String key) =>
    (f[key] as List).cast<Map<String, dynamic>>();

void main() {
  final raw = File('../shared/fixtures/sap_order_state.json').readAsStringSync();
  final fixture = json.decode(raw) as Map<String, dynamic>;

  group('an order', () {
    for (final c in _cases(fixture, 'order_status')) {
      test(c['why'] as String, () {
        expect(
          productionStatusFromSap(SapOrderState(
            salesOrder: c['sap_order'] as String,
            invoice: c['invoice'] as String,
          )),
          c['expect'],
        );
      });
    }
  });

  group('one line of an order', () {
    for (final c in _cases(fixture, 'line_status')) {
      test(c['why'] as String, () {
        expect(
          lineStatusFromSap(
            SapLineState(invoice: c['line_invoice'] as String),
            order: SapOrderState(salesOrder: c['order_sap'] as String),
          ),
          c['expect'],
        );
      });
    }

    test('without its order, an uninvoiced line cannot claim to be in SAP', () {
      expect(lineStatusFromSap(const SapLineState()), kSapNotStarted);
    });
  });

  group('what an order list shows', () {
    for (final c in _cases(fixture, 'order_progress')) {
      test(c['why'] as String, () {
        expect(
          orderProgress(
            SapOrderState(
              salesOrder: c['sap_order'] as String,
              invoice: c['invoice'] as String,
            ),
            c['stored'],
          ),
          c['expect'],
        );
      });
    }

    test('a missing stored status is Not Started, not the string null', () {
      expect(orderProgress(const SapOrderState(), null), kSapNotStarted);
    });
  });

  group('an order rolls up from its lines', () {
    for (final c in _cases(fixture, 'order_rolls_up_from_lines')) {
      test(c['why'] as String, () {
        final lines = (c['line_invoices'] as List)
            .cast<String>()
            .map((i) => SapLineState(invoice: i))
            .toList();
        expect(
          orderStatusFromLines(
            lines,
            SapOrderState(
              salesOrder: c['order_sap'] as String,
              invoice: c['order_invoice'] as String?,
            ),
          ),
          c['expect'],
        );
      });
    }
  });

  group('the one line a rep reads', () {
    for (final c in _cases(fixture, 'summary')) {
      test(c['why'] as String, () {
        final order = (c['order'] as Map).cast<String, dynamic>();
        expect(sapSummary(SapOrderState.fromOrder(order)), c['expect']);
      });
    }
  });

  group('reading it off a Sales Order document', () {
    test('every field maps', () {
      final s = SapOrderState.fromOrder({
        'custom_sap_sales_order': '412',
        'custom_sap_sales_order_status': 'bost_Open',
        'custom_sap_invoice': 'INV-900',
        'custom_sap_invoice_date': '2026-09-24',
        'custom_sap_synced_at': '2026-09-24 10:00:00',
        'custom_sap_sync_error': '',
      });
      expect(s.salesOrder, '412');
      expect(s.salesOrderStatus, 'bost_Open');
      expect(s.invoice, 'INV-900');
      expect(s.invoiceDate, '2026-09-24');
      expect(productionStatusFromSap(s), kSapDispatched);
    });

    test('an old production stage is not read at all', () {
      // Orders from before 24 Sep 2026 can still carry a stage. Nothing may
      // act on it: a stage says nothing about whether the order has gone.
      final s = SapOrderState.fromOrder({
        'custom_sap_sales_order': '412',
        'custom_sap_production_stage': 'Closed',
        'custom_sap_production_order': '4482',
      });
      expect(productionStatusFromSap(s), kSapPushed);
    });

    test('an old delivery is not dispatch', () {
      // The floor delivers before it invoices; only the invoice is dispatch.
      final s = SapOrderState.fromOrder({
        'custom_sap_sales_order': '412',
        'custom_sap_delivery_order': '4',
      });
      expect(productionStatusFromSap(s), kSapPushed);
    });

    test('reading line state off an items row', () {
      final l = SapLineState.fromLine({
        'custom_sap_invoice': 'INV-900',
        'custom_sap_invoice_date': '2026-09-24',
        'custom_sap_production_stage': 'Curing',
      });
      expect(l.hasSap, isTrue);
      expect(l.invoiceDate, '2026-09-24');
    });

    test('an items row with only a stale stage says nothing of its own', () {
      final l = SapLineState.fromLine({'custom_sap_production_stage': 'Curing'});
      expect(l.hasSap, isFalse);
    });
  });

  group('the mistakes this mapping exists to prevent', () {
    test("Frappe's string 'null' is not an invoice", () {
      expect(
          lineStatusFromSap(const SapLineState(invoice: 'null'),
              order: const SapOrderState(salesOrder: '412')),
          kSapPushed);
    });

    test('not yet in SAP is not the same as failed', () {
      // Both have no SAP number; only the sync error means something is wrong.
      const waiting = SapOrderState();
      const failed = SapOrderState(syncError: 'SAP refused the order');
      expect(productionStatusFromSap(waiting), kSapNotStarted);
      expect(productionStatusFromSap(failed), kSapNotStarted);
      expect(waiting.syncError, isNull);
      expect(failed.syncError, isNotEmpty);
    });
  });

  group('staleness', () {
    final now = DateTime.parse('2026-09-12T10:00:00');
    test('an order SAP never took is waiting, not stale', () {
      expect(sapStale(const SapOrderState(), now), isFalse);
    });
    test('in SAP but never reconciled is stale', () {
      expect(sapStale(const SapOrderState(salesOrder: 'SO-1'), now), isTrue);
    });
    test('reconciled this morning is not stale', () {
      expect(
          sapStale(
              const SapOrderState(
                  salesOrder: 'SO-1', syncedAt: '2026-09-12T08:00:00'),
              now),
          isFalse);
    });
  });

  group('An order SAP has cancelled', () {
    // The failure this closes, found 18 September 2026: SAP order 399 had been
    // cancelled and the sync had recorded bost_Cancelled on the ERPNext order
    // correctly, for days. Nothing read it, so it still showed as approved.
    for (final c in _cases(fixture, 'cancelled_orders')) {
      test(c['why'] as String, () {
        final order = (c['order'] as Map).cast<String, dynamic>();
        expect(cancelledInSap(SapOrderState.fromOrder(order)), c['expect_cancelled']);
      });
    }

    test('cancellation outranks the approval status', () {
      const s = SapOrderState(salesOrderStatus: 'bost_Cancelled');
      expect(orderApprovalLabel('PO Approved - Ready for SAP', s), 'Cancelled in SAP');
    });

    test('an open order keeps its approval label', () {
      const s = SapOrderState(salesOrderStatus: 'bost_Open');
      expect(orderApprovalLabel('PO Approved - Ready for SAP', s), 'Approved');
    });
  });
}
